pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -0,0 +1,211 @@
+		isTransparentProxy = proto.BoolTrue
+	}
+
+	tcpProxyFilter := newTcpProxyListenerFilter(env, node)
+	actualWildcard, _ := getActualWildcardAndLocalHost(node)
+	// add an extra listener that binds to the port that is the recipient of the iptables redirect
+	builder.virtualListener = &xdsapi.Listener{
+		Name:           VirtualListenerName,
+		Address:        util.BuildAddress(actualWildcard, uint32(env.Mesh.ProxyListenPort)),
+		Transparent:    isTransparentProxy,
+		UseOriginalDst: proto.BoolTrue,
+		FilterChains: []listener.FilterChain{
+			{
+				Filters: []listener.Filter{*tcpProxyFilter},
+			},
+		},
+	}
+	return builder
+}
+
+func (builder *ListenerBuilder) buildInboundSplitListener(env *model.Environment, node *model.Proxy) *ListenerBuilder {
+	shouldSplitInOutBound := node.IsInboundOutboundListenerSplitEnabled()
+	if !shouldSplitInOutBound {
+		log.Debugf("Inbound and outbound listeners are united in for node %s", node.ID)
+		return builder
+	}
+	var isTransparentProxy *google_protobuf.BoolValue
+	if node.GetInterceptionMode() == model.InterceptionTproxy {
+		isTransparentProxy = proto.BoolTrue
+	}
+
+	tcpProxyFilter := newTcpProxyListenerFilter(env, node)
+	actualWildcard, _ := getActualWildcardAndLocalHost(node)
+	// add an extra listener that binds to the port that is the recipient of the iptables redirect
+	builder.virtualInboundListener = &xdsapi.Listener{
+		Name:           VirtualInboundListenerName,
+		Address:        util.BuildAddress(actualWildcard, ProxyInboundListenPort),
+		Transparent:    isTransparentProxy,
+		UseOriginalDst: proto.BoolTrue,
+		FilterChains: []listener.FilterChain{
+			{
+				Filters: []listener.Filter{*tcpProxyFilter},
+			},
+		},
+	}
+	return builder
+}
+
+func (builder *ListenerBuilder) getListeners() []*xdsapi.Listener {
+	nInbound, nOutbound, nManagement := len(builder.inboundListeners), len(builder.outboundListeners), len(builder.managementListeners)
+	nVirtual, nVirtualInbound := 0, 0
+	if builder.virtualListener != nil {
+		nVirtual = 1
+	}
+	if builder.virtualInboundListener != nil {
+		nVirtualInbound = 1
+	}
+	nListener := nInbound + nOutbound + nManagement + nVirtual + nVirtualInbound
+
+	listeners := make([]*xdsapi.Listener, 0, nListener)
+	listeners = append(listeners, builder.inboundListeners...)
+	listeners = append(listeners, builder.outboundListeners...)
+	listeners = append(listeners, builder.managementListeners...)
+	if builder.virtualListener != nil {
+		listeners = append(listeners, builder.virtualListener)
+	}
+	if builder.virtualInboundListener != nil {
+		listeners = append(listeners, builder.virtualInboundListener)
+	}
+
+	log.Debugf("Build %d listeners for node %s including %d inbound, %d outbound, %d management, %d virtual and %d virtual inbound listeners",
+		nListener,
+		builder.node.ID,
+		nInbound, nOutbound, nManagement,
+		nVirtual, nVirtualInbound)
+	return listeners
+}
+
+func newTcpProxyListenerFilter(env *model.Environment, node *model.Proxy) *listener.Filter {
```

Comment:
you need separate functions for inbound and outbound. And on inbound path, it should always be passthrough

Reply:
Sorry I don't get your point. 
- Inbound and outbound are united before this PR, so the previous inbound behavior is kind of undefined, right?
- What is the purpose of inbound listener setting passthrough when it is sidecar and outhbound policy is not ALLOW_ANY? Can you elaborate?

Reply:
This change is only for the inbound path. 

Outbound path - 1.3.

Reply:
And yes, passthrough must be set for inbound, as Shriram said.

Reply:
With the recent fix the listener_builder is used only if the feature is enabled.
When the feature is not enabled, use the old code path